File: backend/app/execution/schemas/execution.py
from pydantic import BaseModel, Field
from uuid import UUID

class ExecutionRequest(BaseModel):
    project_id: UUID
    file_id: UUID
    language: str
    # Limit stdin to 64KB to prevent memory exhaustion from large input payloads
    input: str = Field(default="", max_length=65536)

class ExecutionResponse(BaseModel):
    language: str
    compile_time_ms: int
    execution_time_ms: int
    memory_used_kb: int
    exit_code: int
    output: str
    status: str

class StopExecutionRequest(BaseModel):
    container_id: str

File: app/lib/api/execution.ts
import { fetchApi } from '../api';

export interface ExecutionRequest {
  project_id: string;
  file_id: string;
  language: string;
  input?: string;
}

export interface ExecutionResponse {
  language: string;
  compile_time_ms: number;
  execution_time_ms: number;
  memory_used_kb: number;
  exit_code: number;
  output: string;
  status: string;
}

export const executionApi = {
  run: async (request: ExecutionRequest): Promise<ExecutionResponse> => {
    const response = await fetchApi('/execution/run', {
      method: 'POST',
      body: JSON.stringify(request)
    });
    return response.data;
  },
  
  stop: async (container_id: string): Promise<void> => {
    await fetchApi(`/execution/run/stop?container_id=${container_id}`, { method: 'POST' });
  }
};
